packages/core/src/sync-realtime/index.ts

```diff
@@ -571,44 +628,74 @@ export const createRealtimeSync = (
-      let isLogMatched = false;
+      let isMatched = false;
 
       for (const filter of logFilters) {
         if (isLogFilterMatched({ filter, block, log })) {
-          isLogMatched = true;
+          // TODO: includeTransactionReceipt
+          // return early if includeTransactionReceipt, otherwise continue checking filters
           requiredTransactions.add(log.transactionHash);
-          if (filter.includeTransactionReceipts) {
-            requiredTransactionReceipts.add(log.transactionHash);
-          }
+          isMatched = true;
         }
       }
 
-      return isLogMatched;
+      return isMatched;
     });
 
-    // Remove call traces that don't match a filter, recording required transactions
-    callTraces = callTraces.filter((callTrace) => {
-      let isCallTraceMatched = false;
-      for (const filter of callTraceFilters) {
-        if (isCallTraceFilterMatched({ filter, block, callTrace })) {
-          isCallTraceMatched = true;
-          requiredTransactions.add(callTrace.transactionHash);
-          if (filter.includeTransactionReceipts) {
-            requiredTransactionReceipts.add(callTrace.transactionHash);
-          }
+    // Initial weak trace filtering before full filtering with factory addresses in handleBlock
+    traces = traces.filter((trace) => {
+      let isMatched = false;
+      for (const filter of transferFilters) {
+        if (
+          isTransferFilterMatched({
+            filter,
+            block: { number: block.number },
+            trace: trace.trace,
+          })
+        ) {
+          // TODO: includeTransactionReceipt
+          // return early if includeTransactionReceipt, otherwise continue checking filters
+          requiredTransactions.add(trace.transactionHash);
+          isMatched = true;
+        }
+      }
+
+      for (const filter of traceFilters) {
+        if (
+          isTraceFilterMatched({
+            filter,
+            block: { number: block.number },
+            trace: trace.trace,
+          })
+        ) {
+          // TODO: includeTransactionReceipt
+          // return early if includeTransactionReceipt, otherwise continue checking filters
+          requiredTransactions.add(trace.transactionHash);
+          isMatched = true;
         }
       }
 
-      return isCallTraceMatched;
+      return isMatched;
     });
 
     ////////
     // Transactions
     ////////
 
-    const transactions = block.transactions.filter(({ hash }) =>
-      requiredTransactions.has(hash),
-    );
+    const transactions = block.transactions.filter((transaction) => {
+      let isMatched = requiredTransactions.has(transaction.hash);
+      for (const filter of transactionFilters) {
+        if (isTransactionFilterMatched({ filter, block, transaction })) {
+          if (filter.includeReverted === false) {
+            // or TODO: includeTransactionReceipt
+            requiredTransactionReceipts.add(transaction.hash);
+            return true;
```

Comment:
Are you sure we want to `return true` here?

Reply:
Yes, since we have matched the transaction filter and we have also added it to the required transactionReceipts, we can return early.